### PR TITLE
Add support for snmp v2 traps

### PR DESCRIPTION
**Pull request by mico-micic**
Extend the existing SNMP sub-agent with support for SNMP v2 traps.

How it is implemented:
The SNMP sub-agent reads all available streams and subscribes to all of them by calling the corresponding backend RCP. Received events are converted to SNMP v2 traps including value-bindings (if available) and sent by net-snmp API.

### CODE REVIEW

File: apps/snmp/snmp_stream.c [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+
+    while ((xncont = xml_child_each(cxparent, xncont, CX_ELMNT)) != NULL){
+        char *node_name = xml_name(xncont);
+        if ((ychild = yang_find(ys, 0, node_name)) != NULL) {
+            switch (yang_keyword_get(ychild)){
+                case Y_LEAF: /* only Y_LEAF type is supported here */
+                    if (get_oid_for_yang_node(ychild, objid, &objidlen) < 0)
+                        goto done;
+                    if ((body = xml_body(xncont)) != NULL){
+                        if ((ret = type_xml2snmp_pre(body, ychild, &xmlstr)) < 0)
```
Comment (olofhagsand) [inline apps/snmp/snmp_stream.c:153]: Memory leak: xmlstr needs to be free:d before a new loop instance. It is freed at the end of the function, but if the while loop iterates more than once the previous value of xmlstr is overwritten.
(tip: run make memsnmp in docker/test)

File: apps/snmp/snmp_stream.c [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+ done:
+    if (streams){
+        for (int i = 0; i < streams_num; i++)
+            free(streams[i]);
+        free(streams);
+    }
+    
+    clixon_debug(CLIXON_DBG_SNMP, "retval: %d", retval);
+    return retval;
+}
```
Comment (olofhagsand) [inline apps/snmp/snmp_stream.c:476]: No newline at end of file

File: docker/test/Dockerfile.native
Comment (olofhagsand) [inline docker/test/Dockerfile.native:1]: Please add this to Dockerfile and Dockerfile.fcgi as well. For non-restconf and FCGI restconf snmp tests.

File: apps/snmp/snmp_stream.c [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+#include "snmp_lib.h"
+
+/*
+ * SNMP v2 notification OID
+ */
+static oid notificationOid[] = {1, 3, 6, 1, 6, 3, 1, 1, 4, 1, 0};
+
+struct stream_socket{
+    qelem_t    sc_q;   /* queue header */
+    int        socket; /* socket */
```
Comment (olofhagsand) [inline apps/snmp/snmp_stream.c:65]: Style: please use prefixes in socket (sc_socket) as well.

File: apps/snmp/snmp_stream.c [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+/*! Read smiv2:oid for given YANG node
+ *
+ * @param[in]  ys       YANG node
+ * @param[out] objid    oid
+ * @param[out] objidlen oid length
+ * @retval     0  OK
+ * @retval    -1  Error
+ */
+int
+get_oid_for_yang_node(yang_stmt *ys, oid *objid, size_t *objidlen)
```
Comment (olofhagsand) [inline apps/snmp/snmp_stream.c:80]: Style: please put each parameter on own line as described in 
https://github.com/clicon/clixon/blob/master/CONTRIBUTING.md#function-declarations
2: ... each parameter on own line
Also at all other places where formal parameters are declared.

File: apps/snmp/snmp_stream.c [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+    return retval;
+}
+
+/*! Add snmp trap v2 oid to the variable list and bind the given notification oid as value.
+ * 
+ * @param[in]  var_list   Variable list
+ * @param[in]  notify_oid OID of the notification to set as value
+ * @param[in]  oid_len    Length of notify_oid
+ */
+void
```
Comment (olofhagsand) [inline apps/snmp/snmp_stream.c:108]: Style: trailing space after void. Please remove all trailing spaces in file

File: apps/snmp/snmp_stream.c [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+ * @param[in]  h  Clixon handle
+ * @retval     0  OK
+ * @retval    -1  Error
+ * @see snmp_stream_subscribe
+ */
+int
+clixon_snmp_stream_init(clixon_handle h)
+{
+    int     retval = -1;
+    struct stream_socket *stream_socket;
```
Comment (olofhagsand) [inline apps/snmp/snmp_stream.c:437]: Style: please indent variables to same level

**Review by olofhagsand — COMMENTED** [review]
Nice PR, thanks.
I cannot really judge the functionality in itself, but the code is clear and the regression test seems sane and passes.
I would also like you to please make a PR to the documentation here https://github.com/clicon/clixon-docs/blob/master/snmp.rst
Altogether a nice contribution, thanks.

**Comment by mico-micic** [comment]
Added PR for documentation: https://github.com/clicon/clixon-docs/pull/7
I think the PR is ready for final review and merge.

**Pull request merged**